### PR TITLE
add validate by jwt token for agents

### PR DESCRIPTION
**Pull request by BoTranVan**

### CODE REVIEW

File: auth/token.go [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+// along with this program.  If not, see <https://www.gnu.org/licenses/>
+
+package auth
+
+import (
+	"io/ioutil"
+	"log"
+	"os"
+
+	"git.paas.vn/OpenStack-Infra/bizfly-agent/config"
```
Comment (cuonglm) [inline auth/token.go:25]: Please do group import:

 - std
 - third party
 - our local modules

File: auth/token.go [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+	"github.com/spf13/viper"
+)
+
+var (
+	cfg                     config.Configurations
+	defaultMetadataEndpoint = cfg.AuthServer.DefaultMetadataEndpoint
+)
+
+// Token is exported
```
Comment (cuonglm) [inline auth/token.go:34]: This comment is meaningless, please add a more meaningful comment or make it a template (using [gocmt](https://github.com/cuonglm/gocmt)).

File: auth/token.go [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+	"io/ioutil"
+	"log"
+	"os"
+
+	"git.paas.vn/OpenStack-Infra/bizfly-agent/config"
+	"github.com/spf13/viper"
+)
+
+var (
+	cfg                     config.Configurations
```
Comment (cuonglm) [inline auth/token.go:30]: Please run `gofmt`.

File: auth/token.go [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+	defaultMetadataEndpoint = cfg.AuthServer.DefaultMetadataEndpoint
+)
+
+// Token is exported
+type Token struct {
+}
+
+// SaveToken is exported
+func (t *Token) SaveToken(token string) {
+	viper.Unmarshal(&cfg)
```
Comment (cuonglm) [inline auth/token.go:40]: This is done in every functions in this file. Let do it in `func init() {}`

File: auth/token.go [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	log.Println("Saving auth token")
+	_, err = file.WriteString(token)
+	if err != nil {
+		log.Fatal(err)
+	}
+	file.Close()
```
Comment (cuonglm) [inline auth/token.go:53]: When writing to file, you have to explicitly check for error return by `file.Close()`.

File: auth/token.go [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+// Token is exported
+type Token struct {
+}
+
+// SaveToken is exported
+func (t *Token) SaveToken(token string) {
+	viper.Unmarshal(&cfg)
+
+	// log.Println(cfg.ConfigDir)
+	file, err := os.Create(cfg.ConfigDir + "/auth_token")
```
Comment (cuonglm) [inline auth/token.go:43]: Use `filepath.Join` instead.

File: client/client.go [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+
+	"git.paas.vn/OpenStack-Infra/bizfly-agent/auth"
```
Comment (cuonglm) [inline client/client.go:29]: Please do group import, like comment above.

File: client/client.go [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+	"github.com/spf13/viper"
+)
+
+var (
+	cfg                     config.Configurations
+	defaultMetadataEndpoint string
+	authToken               auth.Token
+)
+
+// Client is exported
```
Comment (cuonglm) [inline client/client.go:40]: Same as comment above.

File: client/client.go [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+type Client struct {
+	httpClient       *http.Client
+	metadataEndpoint string
+	authToken        string
+}
+
+// NewHTTPClient is exported
```
Comment (cuonglm) [inline client/client.go:47]: Same as comment above.

File: client/client.go [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+func NewHTTPClient() *Client {
+	return &Client{
+		httpClient: http.DefaultClient,
+	}
+}
+
+// AuthToken is exported
```
Comment (cuonglm) [inline client/client.go:54]: Same as comment above.

File: client/client.go [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+func (c *Client) AuthToken() (string, error) {
+	viper.Unmarshal(&cfg)
```
Comment (cuonglm) [inline client/client.go:56]: Same as comment in package `auth`.

Maybe you should create a package `config`, and just do loading config once there.

File: main.go [hunk 12]
```diff
@@ -1,46 +1,68 @@
...
+	pusher := push.New(pushGatewayAddress, "bizfly-agent").
 		Client(httpClient).
-		Grouping("instance_id", instanceID).
+		Grouping("instance_id", cfg.Agent.ID).
+		Grouping("hostname", cfg.Agent.Hostname).
+		Grouping("instance", cfg.Agent.Hostname).
 		Collector(nc)
 
 	if err := pusher.Push(); err != nil {
-		log.Errorf("failed to make initial push to push gateway: %s\n", err.Error())
```
Comment (cuonglm) [inline main.go:38]: Why change this from `Errorf` to `Fatalf`?

IIRC, we don't abort the agent if pushing fails cc @greatbn.

Maybe adding some kind of backoff here?

File: auth/token.go (hunk 3 above)
Comment (BoTranVan) [reply auth/token.go:32]: I use gofmt but format code still same it.

File: auth/token.go (hunk 4 above)
Comment (BoTranVan) [reply auth/token.go:43]: No, if I put it in `func init() {}` agent will not run, beacause var ConfigDir is ""

File: client/client.go (hunk 11 above)
Comment (BoTranVan) [reply client/client.go:57]: I don't know do that. Can you help me?

File: auth/token.go (hunk 4 above)
Comment (cuonglm) [reply auth/token.go]: @BoTranVan See my latest commits

File: client/client.go (hunk 11 above)
Comment (cuonglm) [reply client/client.go]: @BoTranVan See my latest commits

File: main.go (hunk 12 above)
Comment (greatbn) [reply main.go:38]: Yes. We should retry to push metric to Metrics Server.

File: bizfly-agent.yaml [hunk 13]
```diff
@@ -0,0 +1,14 @@
+# Where save auth token
+configDir: .
+
+agent:
+  id: 5ec5eef8a0a0be0018364823
+  name: gooogle-node-apjxd08l
+  hostname: gooogle-node-apjxd08l
+
+authServer:
+  defaultMetadataEndpoint: http://192.168.53.89:8009
```
Comment (greatbn) [inline bizfly-agent.yaml:10]: @BoTranVan `AuthServer` and `Metadata Endpoint`. It does not make sense.

File: client/client.go [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+	return c
+}
+
+// AuthToken get, save and set auth token
+func (c *Client) AuthToken() (string, error) {
+	if config.Config.AuthServer.DefaultEndpoint == "" {
+		prol.Fatalln("Default Metadata Endpoint is required")
+	}
+	c.metadataEndpoint = config.Config.AuthServer.DefaultEndpoint
+	resp, err := c.Get(fmt.Sprintf("%s/agent_tokens?agent_id=%s", c.metadataEndpoint, config.Config.Agent.ID))
```
Comment (greatbn) [inline client/client.go:59]: Get token without authentication information? 

In this case, I know your `agent_id`, I can push fake data.

File: client/client.go [hunk 15]
```diff
@@ -53,15 +54,30 @@ func NewHTTPClient() *Client {
...
 	if config.Config.AuthServer.DefaultEndpoint == "" {
-		prol.Fatalln("Default Metadata Endpoint is required")
+		prol.Fatalln("Default Endpoint is required")
 	}
-	c.metadataEndpoint = config.Config.AuthServer.DefaultEndpoint
-	resp, err := c.Get(fmt.Sprintf("%s/agent_tokens?agent_id=%s", c.metadataEndpoint, config.Config.Agent.ID))
+	c.defaultEndpoint = config.Config.AuthServer.DefaultEndpoint
+	req, err := http.NewRequest("GET", fmt.Sprintf("%s/agent_tokens?agent_id=%s", c.defaultEndpoint, config.Config.Agent.ID), nil)
+	if err != nil {
+		log.Fatal("Error reading request. ", err)
```
Comment (cuonglm) [inline client/client.go:62]: Why not use `prol`?

File: client/client.go [hunk 16]
```diff
@@ -59,7 +58,7 @@ func (c *Client) AuthToken() (string, error) {
 	c.defaultEndpoint = config.Config.AuthServer.DefaultEndpoint
 	req, err := http.NewRequest("GET", fmt.Sprintf("%s/agent_tokens?agent_id=%s", c.defaultEndpoint, config.Config.Agent.ID), nil)
 	if err != nil {
-		log.Fatal("Error reading request. ", err)
+		prol.Error("Error reading request. ", err)
```
Comment (cuonglm) [inline client/client.go:61]: Missing `return "", err`

Same for other places.

File: client/client.go [hunk 17]
```diff
@@ -0,0 +1,139 @@
...
+	if c.authToken != nil {
+		_ = c.authToken.SaveToken(tokenStr)
+	}
+
+	c.token = tokenStr
+	return tokenStr, nil
+}
+
+// Get ...
+func (c *Client) Get(url string) ([]byte, error) {
```
Comment (greatbn) [inline client/client.go:89]: This function is not used, right ?

Comment (cuonglm) [reply client/client.go:89]: It's used in `c.AuthToken` above.

Comment (BoTranVan) [reply client/client.go]: No, it's no used.

**Pull request merged**